% ************************** Thesis Abstract *****************************
% Use `abstract' as an option in the document class to print only the titlepage and the abstract.
\begin{abstract}

Variation in bacterial genomes has important implications for global public health. For example, the ability to identify transmission events and predict antimicrobial drug resistance (AMR) depends on such differences. A single reference genome is often used as a template to find variation. However, bacterial species have a pan-genome, whereby some genes are common to all and others only appear once. Genome graphs are a novel approach to genomic analysis that suit pan-genomes. They use a collection of known genetic variation instead of a single reference. Previous genome graph work has developed methods for representing bacterial genomes as a set of unordered graphs, where each graph represents a locus such as a gene. One limitation though is it can only describe genomes as a mosaic of known variation. Finding novel variation is central to many applications, such as those mentioned earlier. Therefore, in the first chapter, we extend this genome graph method to allow de novo variant discovery and develop a process for evaluating graph-based variant calls.

Whole-genome sequencing (WGS) is used by many public health authorities to surveil bacterial pathogens. Mycobacterium tuberculosis - the cause of tuberculosis (TB) - is one species of particular public health interest as it kills more people each year than any other pathogen. Some countries use Illumina WGS to assess TB patients for transmission clusters and AMR, although this is not ideal in some high-burden settings, which often have reduced resources. Nanopore sequencing presents a potential boon for providing the same information as Illumina with greater speed and lower resource requirements. However, there has been no detailed study of Nanoporeâ€™s ability to provide this information for TB, which is the aim of the remaining three chapters. First, we show Nanopore data can identify transmission clusters consistent with those from Illumina and investigate the potential for genome graphs to help with this task. Second, we demonstrate that Nanopore provides AMR predictions concordant with Illumina and introduce a framework for using genome graphs to improve current methods. Lastly, we train a Nanopore basecalling model for TB and show that it provides a considerable improvement in read- and assembly-level accuracy.

Taken together, this work allows genome graphs to better interrogate variation in bacterial genomes and opens the door for Nanopore-based TB public health applications.

\end{abstract}

