%!TEX root = ../thesis.tex
%*******************************************************************************
%******************************   Fifth Chapter   ***************************
%*******************************************************************************
\chapter*{Discussion}
\addcontentsline{toc}{chapter}{Discussion} \markboth{Discussion}{}
\label{chap:conclusion}
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

% In this case we are not going to ask for any changes other than to add a concluding discussion section. This should address the wider implications of the methods and technologies described in the thesis, not only in the context of current practice but also the potential for new approaches based on an understanding of the biology and genetics of TB and drug resistance. We would like you to take a step back from the focus on current monitoring and clinical methodologies, and discuss how the deeper aims of the field might be advanced using the technologies and computational approaches your work is based on.
% You have 3 months to complete this, but we expect that it should take much less than that - a few pages is what we have in mind. Hopefully it should be a useful exercise to go through in terms of where you see your research going in future.


This thesis has focused on novel methodologies (genome graphs) and technology (nanopore sequencing) as applied to bacterial genomics and particularly \mtb{}. In this concluding section, we will take a step back and outline where we think these can have an impact over the coming years.

\hspace{0.5cm}

% 1. genome graphs for recombining bacteria: suppose you want to understand the genetic basis of some phenotype, and that you want to do this by a simple GWAS design. You have cases and controls and you want to perform associations on all the SNPs you can find in the core genome. The larger your cohort is, the smaller the core genome is. So the more data you have, the less effective the core-genome-SNPs approach is. This is a fundamental problem. People have tried getting away from this by using kmers, but there are bloody loads of them - multiple testing challenge, and even if one is a hit (highly correlated with the phenotype),you then have to map it to the pan-genome and figure out what it is. It's a pain - hard to interpret.

%  Genome graphs provide a means for genotyping SNPs across the whole cohort, so you can just take a multi-sample VCF that describes (almost)the whole pan-genome. and then do your GWAS/association tests. This really opens things up got geno/pheno in bacteria,and lots of prior datasets can be reanalysed. 

\noindent
Pan-genome reference graphs, as used by Pandora, will eventually facilitate the simultaneous analysis of thousands to tens of thousands of bacterial samples. Such population-scale analyses will open the door for many novel insights into the bacterial pan-genome. One future application is genome-wide association studies (GWAS) in recombining bacteria. An example use case for this would be to understand the genetic basis for a phenotype of interest. Doing this via a simple GWAS design, with cases and controls, one would look for associations between core genome SNPs and the phenotype. However, the larger the cohort, the smaller the core genome becomes (\autoref{fig:pangenome-size}), and thus the less effective the GWAS will be. 

Reference-free alternatives to this approach have been proposed, whereby associations between \kmer{}s and the phenotype are used instead of SNPs \cite{Jaillard2018,Lees2016Sequence}. However, there are \emph{vast} numbers of \kmer{}s in a pan-genome, which creates statistical issues with multiple testing. In addition, \kmer{}s do not represent a biological entity. That is, if one finds a \kmer{} to be significant, its location must be found by mapping to the pan-genome to determine its function.

As shown throughout this thesis, genome graphs provide a means to genotype SNPs across a cohort. Pandora provides a multi-sample VCF that describes not just the core genome but (nearly) the \emph{whole} pan-genome, and the GWAS tests can then be performed on this VCF. This capability will facilitate new insights into genotype-phenotype association in bacteria via analyses of prior datasets and future population-scale ones.


% 2. metagenomes - unit of selection may be many different things - gene, strain, plasmid, transposon. pandora is flexible enough you could focus on any of these and study it. Lots of interedt in dynamics in the microbiome at the moment,  how it responds to drugs etc, or longitudinal studies in chronically infected patients (eg cystic fibrosis)- pandoras provides a nice method you can look at all of these, and the multi-sample VCF is a nice way to look at longitudinal samples and compare easily.

\noindent
The pan-genomic model we have described in this thesis lets go of gene ordering and even the definition of a locus. Many bacterial species have loci in common and can even share these loci. Indeed, some metagenomic applications disregard species definitions entirely and focus purely on loci. Furthermore, the unit of selection (loci) could be one of many different classes - e.g., gene, strain, plasmid, transposon. Pandora provides the flexibility to handle all of these definitions. As such, meta-genome graphs are something we believe will provide many fruitful insights. One application we foresee meta-genome graphs enabling is longitudinal studies of the microbiome and its response to drugs in patients with chronic infections - e.g., chronic \textit{Pseudomonas aeruginosa} infections in the lungs of cystic fibrosis patients and its impact on the lung microbiome \cite{Boutin2017Chronic}.

% 3. TB AMR: during this thesis i took a very narrow perspective, trying to show nanopore was "non-inferior" to illumina for some very specific questions: DST and clustering. These are the key pragmatic clinical questions. But more broadly, there are a lot of very interesting biological questions which nanopore data might allow us to look at. First: in regions with very high burden, people can be infected by multiple strains of Mtb, leading to a challenging statistical problem when analysing sequence data - similar to "phasing" in human genomics, but with unknown/complex ploidy. Long nanopore reads definitely help here as they can connect multiple SNP alleles. Caveat: M. tuberculosis has an annoyingly tough cell wall, which means DNA extraction processes have to be slightly brutal, and in the process DNA gets fragmented. It is therefore currently difficuly tyo get very long reads, as are currently possible with human or e coli for example. so woud need to fix this.  This would allow us to look at the dynamics of different strains longiyudinally, see how different mutations compete within host.

\noindent
This thesis took a narrow perspective in showing \ont{}'s "non-inferiority" to Illumina for \mtb{} DST and clustering. However, \ont{} provides unique technological differences that will allow for exploring other very interesting, clinically-focused, biological questions. For one, polyclonal (mixed) infections commonly occur in tuberculosis \cite{Cohen2012} and in most pathogens \cite{Balmer2011}. Detecting such mixed infections from genomic data can be challenging, similar to phasing haplotypes in human genomics, but with unknown or complex ploidy. As long \ont{} reads span multiple informative variants, they help with such phasing problems \cite{Patterson2015WhatsHap}. In addition, we believe genome graphs will allow for determining such mixed infections in partnership with long reads based on support for different paths in a \prg{}. This would allow one to look at the longitudinal dynamics of different strains throughout infection and see how mutations compete within the host during treatment. 

% 4. nanopore: i would think potential for multiomic analyses, where you look at dna, rna and methylation in drug sensitive and resistant strains is v exciting.

Another exciting possibility off the back of the methods and technologies in this thesis is the high-resolution investigation of \ppe{} genes. These genes have been notoriously difficult to probe due to the difficulty in mapping short reads to them. These 160 genes pose many difficulties for short reads due to their high GC content, homology, and repetitive nature. As a result, they are generally excluded from genomic analyses. \ont{} reads promise to improve interrogation resolution as they can span entire (and multiple) \ppe{} genes - leaving much less ambiguity about where a read aligns. Indeed there has been some initial investigation of exactly this \cite{bainomugisa2018}. While this work dramatically improved coverage of these genes, gaps remain. This is where genome graphs are likely to complement \ont{}; by increasing \textit{a priori} knowledge and reducing ambiguity. Added to the fact that these genes have a higher mutation rate than the rest of the genome \cite{bainomugisa2018}, genome graphs are ideally suited to handle the varying complexities of these genes. Increased resolution in the \ppe{} genes promises to improve our knowledge of \mtb{} host-pathogen interactions, virulence, AMR, and other yet-to-be-determined functions.

The prospect of real-time clinical diagnostics is a major attraction of \ont{} sequencing. While this functionality has been shown in theory, it remains to be implemented for \mtb{}. A critical component of this application will be direct-from-sputum \ont{} sequencing, which is still in its infancy. However, significant progress has been made \cite{George2020,Votintseva2017}, and it is only a matter of time until this is routine. In aid of improving direct-from-sputum sequencing, we see \emph{Read Until} \cite{Payne2021,Kovaka2021} as playing an important role. Read Until is an API to the flowcell that allows for artificial depletion/enrichment of a sample (or target regions) by permitting ejection of reads from a pore based on some criteria. Potential applications include rejection of non-\mtb{} reads or enriching AMR-associated or \ppe{} genes.

One final application that will likely advance \mtb{} research and clinical care is epigenetics. \ont{} sequencing inherently captures information about epigenetic modifications, and certain modifications can now be identified in other species \cite{Furlan2021}. Epigenetic modifications have been associated with drug resistance, virulence, and regulation of gene expression profiles in \mtb{} \cite{Phelan2018Methylation,Shell2013,Zhu2016Precision}, but as yet, no exploration of this with \ont{} sequencing has been attempted. Combining this with direct RNA sequencing - another unique capability of \ont{} - the next decade should see a much better understanding of how gene expression and RNA modification influence \mtb{} functions.
