%!TEX root = ../thesis.tex
%*******************************************************************************
%******************************   Fifth Chapter   ***************************
%*******************************************************************************
\chapter*{Conclusion}
\addcontentsline{toc}{chapter}{Conclusion} \markboth{Conclusion}{}
\label{chap:conclusion}
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

This thesis shows that genome graphs and \ont{} sequencing can be combined to provide valuable insight into genomic variation within bacteria. In particular, we have demonstrated how these two techniques can be combined to improve public health applications for \mtb{}.

We began this work by describing a method for novel variation discovery in bacterial genome graphs with Pandora (\autoref{chap:denovo}). We then showcased how Pandora's genome graph approach improves variant recall (Illumina and \ont{}) and precision (\ont{}) compared to traditional linear genome methods.

\autoref{chap:clustering} demonstrated that, for \mtb{}, \ont{} sequencing data produces SNP calls of equal precision to Illumina, with a slight decrease in recall. We then established that \ont{} SNP calls from BCFtools lead to putative transmission clusters consistent with those from Illumina. That is, \ont{}-based clusters do not miss any samples from clustering - no false negatives - and have a small number of (false positive) additional clustered samples. While Pandora was able to provide clusters with no false negatives, the rate of false positives was much higher than BCFtools. However, we outlined some avenues for future work to improve this poor false positive performance; including a different approach to constructing the loci reference graphs. 

In \autoref{chap:dst}, we established the concordance of antimicrobial resistance predictions between Illumina and \ont{} sequencing data with the tool Mykrobe. Predictions from the two technologies were consistent, except for isoniazid, where we found a high false-positive rate in \ont{} data - driven by systematic indel errors. In addition, we implemented a new software program, \drprg{}, which uses Pandora to predict drug resistance. \drprg{} predictions were consistent with Mykrobe for both sequencing technologies, albeit with a slight increase in the false-negative rate for isoniazid. Most importantly, \drprg{} can identify off-catalogue mutations and return unknown predictions for the relevant drug(s). These novel mutations are precise and reduce the number of missed resistance calls.

Finally, we trained a species-specific \ont{} basecalling model that produces more accurate \mtb{} reads and assemblies than the default model (\autoref{chap:tubby}). This improvement in accuracy coincides with a decrease in homopolymer insertions and deletions - two known systematic issues with \ont{} data.

% In this case we are not going to ask for any changes other than to add a concluding discussion section. This should address the wider implications of the methods and technologies described in the thesis, not only in the context of current practice but also the potential for new approaches based on an understanding of the biology and genetics of TB and drug resistance. We would like you to take a step back from the focus on current monitoring and clinical methodologies, and discuss how the deeper aims of the field might be advanced using the technologies and computational approaches your work is based on.
% You have 3 months to complete this, but we expect that it should take much less than that - a few pages is what we have in mind. Hopefully it should be a useful exercise to go through in terms of where you see your research going in future.

\hspace{0.5cm} 

\noindent
The methods and technologies described in this thesis can advance existing approaches and enable new ones. For example, in the context of bacterial genomics in general, pan-genome reference graphs, as used by Pandora, will eventually facilitate the simultaneous analysis of thousands to tens of thousands of samples. Such population-scale analyses will open the door for many novel insights into the bacterial pan-genome. One future application is the fine-grained analysis of the accessory genome at scale. Investigating the accessory genome at this scale, with the resolution Pandora enables, will allow for a greater understanding of the genetic determinants underlying bacterial adaptation to different niches (e.g., host).

% Genome-wide association studies (GWAS) are another application that will benefit from such pan-genome graph methods. GWAS has traditionally been difficult to perform in bacteria due to XXX (reference variability?). Current GWAS approaches, termed pan-GWAS only correlate accessory genes to phenotypic traits. However, we foresee Pandora enabling the correlation of individual variants to traits - a dramatic improvement in scale. 

% \todo[inline]{Are there bacterial GWAS methods that already do SNP-GWAS?}
% \todo[inline]{How/will pandora allow SNP-scale GWAS? What is it that enables this?}
% \todo[inline]{Maha has some TB GWAS papers - check these out}

Looking even further ahead, the pan-genomic model we have described in this thesis - where we let go of gene ordering - lends itself to allowing for the relaxation of species requirements. Many bacterial species have genes in common and can even share genes. Indeed, some metagenomic applications disregard species definitions and focus purely on loci. In an environmental (biomonitoring) context, one may only care about the presence or absence of particular loci such as AMR genes or even specific plasmids. Pandora will facilitate improvements in these analyses with nucleotide-level resolution within these loci of interest. As such, meta-genome graphs are one application we believe will provide many fruitful insights.

While not a current capability, we believe it will be possible to use Pandora to determine the abundance of different paths in a reference graph. This ability will allow the inference of polyclonal (mixed) samples. Not only would this be useful in the environmental example just outlined, but also in clinical applications. Multiple-strain infections commonly occur in tuberculosis \cite{Cohen2012} - and indeed, in most pathogens \cite{Balmer2011}. These mixed infections can become quite insidious when a minor clone has resistance to a drug that the major clone is susceptible to. If a patient is put on a treatment regimen without the knowledge of the minor clone, it will likely sweep to dominance, placing the patient back at the beginning of treatment. Therefore, we foresee the ability to (1) infer the abundance of mixed strains; and (2) predict the resistance profile of each strain. Long \ont{} reads are likely to aid in this by capturing vital long-range haplotype information.

Another exciting possibility off the back of the methods and technologies in this thesis is the high-resolution investigation of \ppe{} genes. These genes have been notoriously difficult to probe due to the difficulty in mapping short reads to them. These 160 genes pose many difficulties for short reads due to their high GC content, homology, and repetitive nature. As a result, they are generally excluded from genomic analyses. \ont{} reads promise to improve interrogation resolution as they can span entire (and multiple) \ppe{} genes - leaving much less ambiguity about where a read aligns. Indeed there has been some initial investigation of exactly this \cite{bainomugisa2018}. While this work dramatically improved coverage of these genes, gaps remain. This is where genome graphs are likely to complement \ont{}; by increasing \textit{a priori} knowledge and reducing ambiguity. Added to the fact that these genes have a higher mutation rate than the rest of the genome \cite{bainomugisa2018}, genome graphs are ideally suited to handle the varying complexities of these genes. Increased resolution in the \ppe{} genes promises to improve our knowledge of \mtb{} host-pathogen interactions, virulence, AMR, and other yet-to-be-determined functions.

The prospect of real-time clinical diagnostics is a major attraction of \ont{} sequencing. While this functionality has been shown in theory, it remains to be implemented for \mtb{}. We foresee a real-time "revolution" where patients in remote areas (e.g., parts of Madagascar or Papua New Guinea) do not have to wait weeks/months to receive the correct treatment, but we can take the lab to them. A critical component of this application will be direct-from-sputum \ont{} sequencing, which is still in its infancy. However, significant progress has been made \cite{George2020,Votintseva2017}, and it is only a matter of time until this is routine. Effective diagnostics in real-time from sputum will not only be a boon for people in remote settings, but anyone, anywhere. We see a personalised approach, whereby novel variants detected in AMR-associated genes (as shown in \autoref{chap:dst}) during such real-time diagnostics allow ambiguous samples to be sent back to reference labs for phenotypic testing (where possible). In addition to AMR diagnostics, real-time transmission clustering could also help identify where further epidemiological investigation is warranted and allow public health officers to act quickly. This is particularly relevant in remote communities, where such transmission events are more easily investigated. Pairing these types of diagnostics with drone delivery is very futuristic indeed \cite{Bahrainwala2020}. 

Adding to the momentum towards this real-time revolution is the recent release (2019) of Oxford Nanopore Technologies' Flongle adapter. It provides a low-output sequencing solution (2 Gb at 90 USD per Flongle flowcell), which is the lowest set-up cost of any sequencing platform currently available. 

In aid of improving direct-from-sputum sequencing, we see \emph{Read Until} \cite{Payne2021,Kovaka2021} as playing an important role. Read Until is an API to the flowcell that allows for artificial depletion/enrichment of a sample (or target regions) by permitting ejection of reads from a pore based on some criteria. Potential applications include rejection of non-\mtb{} reads or enriching AMR-associated or \ppe{} genes.

One final application that will likely advance \mtb{} research and clinical care is epigenetics. \ont{} sequencing inherently captures information about epigenetic modifications, and certain modifications can now be identified in other species \cite{Furlan2021}. Epigenetic modifications have been associated with drug resistance, virulence, and regulation of gene expression profiles in \mtb{} \cite{Phelan2018Methylation,Shell2013,Zhu2016Precision}, but as yet, no exploration of this with \ont{} sequencing has been attempted. Combining this with direct RNA sequencing - another unique capability of \ont{} - the next decade should see a much better understanding of how gene expression and RNA modification influence \mtb{} functions.
